### PR TITLE
added volume snapshot and clone

### PR DESCRIPTION
**Pull request by geoffrey1330**
1. Added feature for volume snapshot and e2e test
2. Added features for volume clone and e2e test
3. Added Snapshot CRD to helm chart
4. Add snapshot parameter in storageclass

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,7 +3,315 @@
 This repo contains Simplyblock CSI ([Container Storage Interface](https://github.com/container-storage-interface/))
 plugin for Kubernetes.
 
-Simplyblock CSI plugin brings high performance block storage to Kubernetes. It provisions SPDK logical volumes on storage node dynamically and enables Pods to access SPDK storage backend through NVMe-oF .
+SPDK CSI plugin brings SPDK to Kubernetes. It provisions SPDK logical volumes on storage node dynamically
```
Comment (boddumanohar) [inline README.md:6]: the changes here were removed and moved to internal folder as a part of PR: https://github.com/simplyblock-io/spdk-csi/pull/39

Please discard the changes in this file.

File: charts/latest/spdk-csi/templates/controller-rbac.yaml [hunk 2]
```diff
@@ -24,10 +24,10 @@ rules:
   verbs: ["get", "list"]
 - apiGroups: [""]
   resources: ["persistentvolumes"]
-  verbs: ["get", "list", "watch", "create", "delete"]
+  verbs: ["get", "list", "watch", "create", "delete", "patch"]
```
Comment (boddumanohar) [inline charts/latest/spdk-csi/templates/controller-rbac.yaml:27]: question: Why is this additional `patch` permission needed?

File: pkg/spdk/controllerserver.go [hunk 3]
```diff
@@ -263,8 +269,10 @@ func prepareCreateVolumeReq(ctx context.Context, req *csi.CreateVolumeRequest, s
 		MaxRWmBytes: params["qos_rw_mbytes"],
 		MaxRmBytes:  params["qos_r_mbytes"],
 		MaxWmBytes:  params["qos_w_mbytes"],
+		MaxSize:     params["max_size"],
```
Comment (boddumanohar) [inline pkg/spdk/controllerserver.go:272]: the type of `params["max_size"]` is a string. Does the API also requires this value to be a string?

Comment (geoffrey1330) [reply pkg/spdk/controllerserver.go:272]: yes the api converts it back to `int`

File: .github/workflows/linux.yaml [hunk 4]
```diff
@@ -47,12 +47,17 @@ jobs:
...
+        docker push simplyblock/spdkcsi:$TAG-arm64
+
+        docker buildx imagetools create -t simplyblock/spdkcsi:$TAG simplyblock/spdkcsi:$TAG-amd64 simplyblock/spdkcsi:$TAG-arm64
+
+        docker buildx imagetools create -t simplyblock/spdkcsi:$TAG-$SHORT_SHA simplyblock/spdkcsi:$TAG-amd64 simplyblock/spdkcsi:$TAG-arm64
+    
         if [ "${{ github.ref }}" = "refs/heads/master" ]; then
-          docker tag simplyblock/spdkcsi:$TAG simplyblock/spdkcsi:latest
-          docker push simplyblock/spdkcsi:latest
+          docker buildx imagetools create -t simplyblock/spdkcsi:latest simplyblock/spdkcsi:$TAG-amd64 simplyblock/spdkcsi:$TAG-arm64
```
Comment (boddumanohar) [inline .github/workflows/linux.yaml:62]: this changes seems to be already present on the master branch. Could you please check why this is happening.

File: Makefile [hunk 5]
```diff
@@ -137,7 +137,13 @@ image: spdkcsi
 	@if [ -n $(HTTP_PROXY) ]; then \
 		proxy_opt="--build-arg http_proxy=$(HTTP_PROXY) --build-arg https_proxy=$(HTTP_PROXY)"; \
 	fi; \
-	sudo docker build --no-cache -t $(CSI_IMAGE) $$proxy_opt \
+	sudo docker build --no-cache -t $(CSI_IMAGE)-amd64 $$proxy_opt \
```
Comment (boddumanohar) [inline Makefile:140]: same here. Please check because this diff doesn't belong to this PR

**Review by boddumanohar — CHANGES_REQUESTED** [review]

File: charts/README.md [hunk 6]
```diff
@@ -58,7 +58,7 @@ The following table lists the configurable parameters of the latest Simplyblock
 
 | Parameter                              | Description                                                                                                              | Default                                                                 |
 | -------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ | ----------------------------------------------------------------------- |
-| `driverName`                           | alternative driver name                                                                                                  | `csi.spdk.io`                                                           |
+| `driverName`                           | alternative driver name                                                                                                  | `csi.simplyblock.io`                                                           |
```
Comment (boddumanohar) [inline charts/README.md:61]: similarly for `csi.simplyblock.io`

File: pkg/util/nvmf.go [hunk 7]
```diff
@@ -156,12 +176,27 @@ func (node *NodeNVMf) PublishVolume(lvolID string) error {
 	if err != nil {
 		return err
 	}
-
 	klog.V(5).Infof("volume published: %s", lvolID)
 	return nil
 }
 
-// UnpublishVolume unexports a volume through NVMf target
+// func (node *NodeNVMf) isVolumePublished(lvolID string) (bool, error) {
```
Comment (boddumanohar) [inline pkg/util/nvmf.go:183]: remove commented code

File: pkg/util/jsonrpc.go [hunk 8]
```diff
@@ -291,6 +291,17 @@ func (client *RPCClient) getVolumeInfo(lvolID string) (map[string]string, error)
 	}, nil
 }
 
+// func (client *rpcClient) isVolumeCreated(lvolID string) (bool, error) {
```
Comment (boddumanohar) [inline pkg/util/jsonrpc.go:294]: remove commented code

File: pkg/spdk/controllerserver.go [hunk 9]
```diff
@@ -567,6 +588,72 @@ func newControllerServer(d *csicommon.CSIDriver) (*controllerServer, error) {
...
+		klog.Errorf("error creating simplyBlock volume: %v", err)
+		return nil, err
+	}
+	vol.VolumeId = fmt.Sprintf("%s:%s", poolName, volumeID)
+	klog.V(5).Info("successfully Restored Snapshot from Simplyblock with Volume ID: ", vol.GetVolumeId())
+
+	return vol, nil
+}
+
+func (cs *controllerServer) handleVolumeSource(srcVolume *csi.VolumeContentSource_VolumeSource, req *csi.CreateVolumeRequest, poolName string, vol *csi.Volume, sizeMiB int64) (*csi.Volume, error) {
```
Comment (boddumanohar) [inline pkg/spdk/controllerserver.go:622]: Just a minor thing. there seems to be a lot of Info logs here. Can you please check if they are needed. If not, we can move then to debug logs

**Review by boddumanohar — CHANGES_REQUESTED** [review]

File: charts/latest/spdk-csi/templates/setup-snapshot-controller.yaml [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+# It should be installed as part of the base Kubernetes distribution in an appropriate
+# namespace for components implementing base system functionality. For installing with
+# Vanilla Kubernetes, kube-system makes sense for the namespace.
+{{- if .Values.snapshotclass.create -}}
+---
+kind: Deployment
+apiVersion: apps/v1
+metadata:
+  name: snapshot-controller
+  namespace: kube-system
```
Comment (boddumanohar) [inline charts/latest/spdk-csi/templates/setup-snapshot-controller.yaml:13]:
```suggestion
 namespace: {{ .Release.Namespace }}
```

please make sure that the all the controllers are installed in the same namespace

File: charts/latest/spdk-csi/templates/snapshot.storage.k8s.io_volumesnapshotclasses.yaml [hunk 11]
```diff
@@ -0,0 +1,139 @@
+{{- if .Values.snapshotclass.create -}}
```
Comment (boddumanohar) [inline charts/latest/spdk-csi/templates/snapshot.storage.k8s.io_volumesnapshotclasses.yaml:1]: can we have all the crds in a single file?

**Review by boddumanohar — CHANGES_REQUESTED** [review]

File: charts/latest/spdk-csi/templates/snapshot.storage.k8s.io_volumesnapshotclasses.yaml (hunk 11 above)
Comment (geoffrey1330) [reply charts/latest/spdk-csi/templates/snapshot.storage.k8s.io_volumesnapshotclasses.yaml:1]: We can separated with `---`

File: pkg/spdk/controllerserver.go (hunk 9 above)
Comment (geoffrey1330) [reply pkg/spdk/controllerserver.go:622]: Will remove unnecessary info logs

File: pkg/util/jsonrpc.go (hunk 8 above)
Comment (geoffrey1330) [reply pkg/util/jsonrpc.go:294]: Sure

File: charts/latest/spdk-csi/templates/snapshot.storage.k8s.io_volumesnapshotclasses.yaml (hunk 11 above)
Comment (geoffrey1330) [reply charts/latest/spdk-csi/templates/snapshot.storage.k8s.io_volumesnapshotclasses.yaml:1]: But it will make the single file ambiguous

File: deploy/kubernetes/testclone.yaml [hunk 12]
```diff
@@ -1,16 +1,34 @@
...
+  name: spdkcsi-pvc-clone
 spec:
   storageClassName: spdkcsi-sc
   dataSource:
-    name: spdk-snapshot
-    kind: VolumeSnapshot
-    apiGroup: snapshot.storage.k8s.io
+    name: spdkcsi-pvc
+    kind: PersistentVolumeClaim
+    apiGroup: ""
```
[the remark this answers is not in the document]
Comment (geoffrey1330) [reply deploy/kubernetes/testclone.yaml:11]: Referencing core kubernetes resource

File: charts/latest/spdk-csi/templates/setup-snapshot-controller.yaml (hunk 10 above)
Comment (geoffrey1330) [reply charts/latest/spdk-csi/templates/setup-snapshot-controller.yaml]: snapshot-controller needs to be installed in the `kube-system` namespace

**Review by hamdy-simplyblock — APPROVED** [review]
Looks good: jsonrpc.go

**Comment by hamdy-simplyblock** [comment]
e2e is working, approved.

**Review by hamdy-simplyblock — APPROVED** [review]
+1
